File: backend/utils/permissions.py
"""Permission infrastructure for OfficeFlow (extends existing role-based auth).

Roles: super_admin, admin, hr, manager, hd, employee
- super_admin: bypasses ALL permission checks
- hd: bypasses ALL dispatch.* permission checks (Head of Dispatch)
- everyone else: must have explicit permissions[]

Backward-compatible: users without a `permissions` field are treated as having [].
"""
from fastapi import HTTPException, Request
from typing import List, Optional

# ----- Permission code registry (Dispatch module) -----
DISPATCH_PERMISSIONS = {
    "dispatch.dashboard": ["view"],
    "dispatch.schedule": ["view", "create", "edit", "cancel", "delete"],
    "dispatch.clients": ["view", "create", "edit", "delete"],
    "dispatch.vendors": ["view", "create", "edit", "delete"],
    "dispatch.officers": ["view", "create", "edit", "delete"],
    "dispatch.post_sites": ["view", "create", "edit", "delete"],
    "dispatch.confirmation": ["view", "manage", "history"],
    "dispatch.reports": ["view", "export"],
    "dispatch.financial": ["view"],
    "dispatch.billing": ["view"],
    "dispatch.audit": ["view"],
}

# Flat list of all valid permission codes
ALL_PERMISSIONS: List[str] = [
    f"{group}.{action}"
    for group, actions in DISPATCH_PERMISSIONS.items()
    for action in actions
]

# Sensitive Dispatch financial fields — stripped from responses unless perm
FINANCIAL_FIELDS = ("duty_rate", "billing_rate", "work_order_number")


def has_permission(user: dict, code: str) -> bool:
    """Check whether the given user has the given permission code."""
    if not user:
        return False
    role = user.get("role", "employee")
    if role == "super_admin":
        return True
    # HD has full Dispatch access
    if role == "hd" and code.startswith("dispatch."):
        return True
    perms = user.get("permissions") or []
    return code in perms


def user_has_any_dispatch_perm(user: dict) -> bool:
    if not user:
        return False
    role = user.get("role", "employee")
    if role in ("super_admin", "hd"):
        return True
    perms = user.get("permissions") or []
    return any(p.startswith("dispatch.") for p in perms)


def require_permission(user: dict, code: str):
    """Raise 403 if user lacks the given permission."""
    if not has_permission(user, code):
        raise HTTPException(
            status_code=403,
            detail=f"You do not have permission ({code}) to perform this action.",
        )


def strip_financial(doc: dict, user: dict) -> dict:
    """Return a copy of the doc with financial fields removed if user lacks perm."""
    if has_permission(user, "dispatch.financial.view"):
        return doc
    clean = dict(doc)
    for f in FINANCIAL_FIELDS:
        clean.pop(f, None)
    return clean


def validate_permission_codes(codes: Optional[List[str]]) -> List[str]:
    """Filter to only known permission codes."""
    if not codes:
        return []
    return [c for c in codes if c in ALL_PERMISSIONS]

File: frontend/src/lib/permissions.js
// Frontend permission helper — mirrors backend utils/permissions.py
export const PERMISSION_GROUPS = [
  {
    key: 'dashboard', label: 'Dashboard',
    perms: [{ code: 'dispatch.dashboard.view', label: 'View Dashboard' }],
  },
  {
    key: 'schedule', label: 'Schedule',
    perms: [
      { code: 'dispatch.schedule.view', label: 'View Schedule' },
      { code: 'dispatch.schedule.create', label: 'Create Schedule' },
      { code: 'dispatch.schedule.edit', label: 'Edit Schedule' },
      { code: 'dispatch.schedule.cancel', label: 'Cancel Schedule' },
      { code: 'dispatch.schedule.delete', label: 'Delete Schedule' },
    ],
  },
  {
    key: 'clients', label: 'Clients',
    perms: [
      { code: 'dispatch.clients.view', label: 'View' },
      { code: 'dispatch.clients.create', label: 'Create' },
      { code: 'dispatch.clients.edit', label: 'Edit' },
      { code: 'dispatch.clients.delete', label: 'Delete' },
    ],
  },
  {
    key: 'vendors', label: 'Vendors',
    perms: [
      { code: 'dispatch.vendors.view', label: 'View' },
      { code: 'dispatch.vendors.create', label: 'Create' },
      { code: 'dispatch.vendors.edit', label: 'Edit' },
      { code: 'dispatch.vendors.delete', label: 'Delete' },
    ],
  },
  {
    key: 'officers', label: 'Security Officers',
    perms: [
      { code: 'dispatch.officers.view', label: 'View' },
      { code: 'dispatch.officers.create', label: 'Create' },
      { code: 'dispatch.officers.edit', label: 'Edit' },
      { code: 'dispatch.officers.delete', label: 'Delete' },
    ],
  },
  {
    key: 'post_sites', label: 'Post Sites',
    perms: [
      { code: 'dispatch.post_sites.view', label: 'View' },
      { code: 'dispatch.post_sites.create', label: 'Create' },
      { code: 'dispatch.post_sites.edit', label: 'Edit' },
      { code: 'dispatch.post_sites.delete', label: 'Delete' },
    ],
  },
  {
    key: 'confirmation', label: 'Confirmation',
    perms: [
      { code: 'dispatch.confirmation.view', label: 'View' },
      { code: 'dispatch.confirmation.manage', label: 'Manage' },
      { code: 'dispatch.confirmation.history', label: 'View History' },
    ],
  },
  {
    key: 'reports', label: 'Reports',
    perms: [
      { code: 'dispatch.reports.view', label: 'View' },
      { code: 'dispatch.reports.export', label: 'Export' },
    ],
  },
  {
    key: 'financial', label: 'Financial',
    perms: [
      { code: 'dispatch.financial.view', label: 'View Duty Rate / Billing Rate / Work Order' },
      { code: 'dispatch.billing.view', label: 'View Billing Reports' },
    ],
  },
  { key: 'audit', label: 'Audit', perms: [{ code: 'dispatch.audit.view', label: 'View Audit Logs' }] },
];

export const ALL_PERMISSION_CODES = PERMISSION_GROUPS.flatMap((g) => g.perms.map((p) => p.code));

export function hasPermission(user, code) {
  if (!user) return false;
  if (user.role === 'super_admin') return true;
  if (user.role === 'hd' && code.startsWith('dispatch.')) return true;
  return (user.permissions || []).includes(code);
}

export function hasAnyDispatchPerm(user) {
  if (!user) return false;
  if (user.role === 'super_admin' || user.role === 'hd') return true;
  return (user.permissions || []).some((p) => p.startsWith('dispatch.'));
}
